codec-http/src/test/java/io/netty/handler/codec/http/HttpClientUpgradeHandlerTest.java

```diff
@@ -0,0 +1,157 @@
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+public class HttpClientUpgradeHandlerTest {
+
+    class FakeSourceCodec implements HttpClientUpgradeHandler.SourceCodec {
+        public void prepareUpgradeFrom(ChannelHandlerContext ctx) {
+        }
+
+        public void upgradeFrom(ChannelHandlerContext ctx) {
+        }
+    }
+
+    class FakeUpgradeCodec implements HttpClientUpgradeHandler.UpgradeCodec {
+        public CharSequence protocol() {
+            return "fancyhttp";
+        }
+
+        public Collection<CharSequence> setUpgradeHeaders(ChannelHandlerContext ctx, HttpRequest upgradeRequest) {
+            return Collections.emptyList();
+        }
+
+        public void upgradeTo(ChannelHandlerContext ctx, FullHttpResponse upgradeResponse) throws Exception {
+        }
+    }
+
+    class UserEventCatcher extends ChannelInboundHandlerAdapter {
+        private Object evt;
+
+        public Object checkUserEvent() {
+            return evt;
+        }
+
+        @Override
+        public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+            this.evt = evt;
+        }
+    }
+
+    @Test
+    public void testSuccessfulUpgrade() {
+        HttpClientUpgradeHandler.SourceCodec sourceCodec = new FakeSourceCodec();
+        HttpClientUpgradeHandler.UpgradeCodec upgradeCodec = new FakeUpgradeCodec();
+        HttpClientUpgradeHandler handler = new HttpClientUpgradeHandler(sourceCodec, upgradeCodec, 1024);
+        UserEventCatcher catcher = new UserEventCatcher();
+        EmbeddedChannel channel = new EmbeddedChannel(catcher);
+        channel.pipeline().addFirst("upgrade", handler);
+
+        channel.writeOutbound(new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "twitter.com"));
+        HttpRequest request = channel.readOutbound();
+
+        assertEquals(request.headers().size(), 2);
+        assertTrue(request.headers().contains(HttpHeaderNames.UPGRADE, "fancyhttp", false));
+        assertTrue(request.headers().contains("connection", "upgrade", false));
+        assertEquals(catcher.checkUserEvent(), HttpClientUpgradeHandler.UpgradeEvent.UPGRADE_ISSUED);
+
+        HttpResponse upgradeResponse =
+            new DefaultHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.SWITCHING_PROTOCOLS);
+        upgradeResponse.headers().add(HttpHeaderNames.UPGRADE, "fancyhttp");
+        channel.writeInbound(upgradeResponse);
+        channel.writeInbound(LastHttpContent.EMPTY_LAST_CONTENT);
+
+        assertEquals(catcher.checkUserEvent(), HttpClientUpgradeHandler.UpgradeEvent.UPGRADE_SUCCESSFUL);
+        assertNull(channel.pipeline().get("upgrade"));
+
+        channel.writeInbound(new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.OK));
+        HttpResponse response = channel.readInbound();
+        assertEquals(response.status(), HttpResponseStatus.OK);
```

Comment:
ensure you call `channel.finish()` and assert the return value. Also ensure you consume all messages and call `release()` on them.